Tests/MongoSwiftTests/DocumentTests.swift

```diff
@@ -680,4 +682,122 @@ final class DocumentTests: MongoSwiftTestCase {
+
+        // swiftlint:disable nesting
+        struct UUIDWrapper: Codable {
+            let uuid: UUID
+        }
+        // swiftlint:enable nesting
+
+        // randomly generated uuid
+        guard let uuid = UUID(uuidString: "2c380a6c-7bc5-48cb-84a2-b26777a72276") else {
+            throw MongoError.bsonDecodeError(message: "Cant create UUID.")
+        }
+
+        let decoder = BSONDecoder()
+
+        // UUID default decoder expects a string
+        decoder.uuidDecodingStrategy = .deferredToUUID
+        let stringDoc: Document = ["uuid": uuid.description]
+        let badString: Document = ["uuid": "hello"]
+        let deferredStruct = try decoder.decode(UUIDWrapper.self, from: stringDoc)
+        expect(deferredStruct.uuid).to(equal(uuid))
+        expect(try decoder.decode(UUIDWrapper.self, from: badString)).to(throwError())
+
+        decoder.uuidDecodingStrategy = .fromString
+        let stringStruct = try decoder.decode(UUIDWrapper.self, from: stringDoc)
+        expect(stringStruct.uuid).to(equal(uuid))
+        expect(try decoder.decode(UUIDWrapper.self, from: badString)).to(throwError())
+
+        decoder.uuidDecodingStrategy = .fromBinary
+        let uuidt = uuid.uuid
+        let bytes = Data(bytes: [
+            uuidt.0, uuidt.1, uuidt.2, uuidt.3,
+            uuidt.4, uuidt.5, uuidt.6, uuidt.7,
+            uuidt.8, uuidt.9, uuidt.10, uuidt.11,
+            uuidt.12, uuidt.13, uuidt.14, uuidt.15
+        ])
+        let binaryDoc: Document = ["uuid": try Binary(data: bytes, subtype: .uuid)]
+        let binaryStruct = try decoder.decode(UUIDWrapper.self, from: binaryDoc)
+        expect(binaryStruct.uuid).to(equal(uuid))
+
+        let badBinary: Document = ["uuid": try Binary(data: bytes, subtype: .generic)]
+        expect(try decoder.decode(UUIDWrapper.self, from: badBinary)).to(throwError())
+    }
+
+    func testDateCodingStrategies() throws {
+        // TODO: add encoding strategy tests SWIFT-282
+
+        // swiftlint:disable nesting
+        struct DateWrapper: Codable {
+            let date: Date
+        }
+        // swiftlint:enable nesting
+
+        let date = Date(timeIntervalSince1970: 125.0)
+
+        let decoder = BSONDecoder()
+
+        // Default is .bsonDateTime
+        let bsonDate: Document = ["date": date]
+        let defaultStruct = try decoder.decode(DateWrapper.self, from: bsonDate)
+        expect(defaultStruct.date).to(equal(date))
+
+        decoder.dateDecodingStrategy = .bsonDateTime
+        let bsonDateStruct = try decoder.decode(DateWrapper.self, from: bsonDate)
+        expect(bsonDateStruct.date).to(equal(date))
+
+        decoder.dateDecodingStrategy = .millisecondsSince1970
+        let msInt64: Document = ["date": date.msSinceEpoch]
+        let msInt64Struct = try decoder.decode(DateWrapper.self, from: msInt64)
+        expect(msInt64Struct.date).to(equal(date))
+
+        let msDouble: Document = ["date": TimeInterval(date.msSinceEpoch)]
```

Comment:
I think we should just call this `Double` over `TimeInterval` - although it's just a `typealias` and makes no functional difference, the [docs](https://developer.apple.com/documentation/foundation/timeinterval) state that a `TimeInterval` is always in seconds, so it is a little confusing to be using it for milliseconds here.

Reply:
Yep, I have been trying to keep that distinction too but must have missed this one. Good catch, fixed.